### PR TITLE
Added file write benchmarks

### PR DESCRIPTION
**Pull request by BebeSparkelSparkel**
In preparation for #446

### CODE REVIEW

File: benchmarks/haskell/Benchmarks/FileWrite.hs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import Data.Text (StrictText)
+import qualified Data.Text.IO as T
+import Data.Text.Lazy (LazyText)
+import qualified Data.Text.Lazy.IO as LT
+import System.IO (Handle)
+
+benchmark :: StrictText -> LazyText -> Handle -> Handle -> Handle -> Benchmark
+benchmark !st !lt !noBufH !lineBufH !blockBufH = bgroup "FileWrite"
+  [ bgroup "StrictText"
+    [ bench "NoBuffering" $ nfIO $ T.hPutStr noBufH st
```
Comment (Bodigrim) [inline benchmarks/haskell/Benchmarks/FileWrite.hs:24]: In general it's better to use `nfAppIO`, and I suggest we do so in new code.

File: benchmarks/haskell/Benchmarks.hs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE CPP #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE NumericUnderscores #-}
```
Comment (Bodigrim) [inline benchmarks/haskell/Benchmarks.hs:3]: This fails with old GHCs.

File: benchmarks/haskell/Benchmarks/FileWrite.hs (hunk 1 above)
Comment (BebeSparkelSparkel) [reply benchmarks/haskell/Benchmarks/FileWrite.hs]: I think I got the text data to normal form but I am not  familiar with nfAppIO so please check. Thanks

File: benchmarks/haskell/Benchmarks.hs (hunk 2 above)
Comment (BebeSparkelSparkel) [reply benchmarks/haskell/Benchmarks.hs]: I did not realize how new that extension is.

File: benchmarks/haskell/Benchmarks/FileWrite.hs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import Data.Text.Lazy (LazyText)
+import qualified Data.Text.IO.Utf8 as Utf8
+import Data.Bifunctor (first)
+
+mkFileWriteBenchmarks :: IO (Handle, IO ()) -> IO (Benchmark, IO ())
+mkFileWriteBenchmarks mkSinkNRemove = do
+  let writeDate = LT.cycle $ fromString [minBound..maxBound]
+      lengths = [0..5] <> [10,20..100] <> [1000,10000,100000]
+      testGroup :: NFData text => String -> (Handle -> text -> IO ()) -> ((StrictText,LazyText) -> text) -> Newline -> IO (Benchmark, IO ())
+      testGroup groupName hPutStr select nl = do
```
Comment (Bodigrim) [inline benchmarks/haskell/Benchmarks/FileWrite.hs:32]: "This binding for ‘hPutStr’ shadows the existing binding"

File: benchmarks/haskell/Benchmarks/FileWrite.hs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    . foldr (\(b,r) (bs,rs) -> (b:bs,r>>rs)) ([], return ())
+    <$> sequence
+    [ testGroup "Strict hPutStr" T.hPutStr    strict LF
+    , testGroup "Lazy   hPutStr" LT.hPutStr   lazy   LF
+    , testGroup "Strict hPutStr" T.hPutStr    strict CRLF
+    , testGroup "Lazy   hPutStr" LT.hPutStr   lazy   CRLF
+    , testGroup "Utf-8  hPutStr" Utf8.hPutStr strict LF
+    ]
+
+strict = fst
```
Comment (Bodigrim) [inline benchmarks/haskell/Benchmarks/FileWrite.hs:69]: "Top-level binding with no type signature: strict :: (a, b) -> a" and same for `lazy` below.

File: benchmarks/haskell/Benchmarks/FileWrite.hs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        hSetBuffering noBufH NoBuffering
+        hSetNewlineMode noBufH nlm
+        (!lineBufH, lineBufRm) <- mkSinkNRemove
+        hSetBuffering lineBufH LineBuffering
+        hSetNewlineMode lineBufH nlm
+        (!blockBufH, blockBufRm) <- mkSinkNRemove
+        hSetBuffering blockBufH $ BlockBuffering Nothing
+        hSetNewlineMode blockBufH nlm
+
+        return
```
Comment (Bodigrim) [inline benchmarks/haskell/Benchmarks/FileWrite.hs:44]: Let's stick to `pure` instead of `return` in new code.

**Review by Bodigrim — COMMENTED** [review]
Could you please fix CI for GHC 8.2?

**Review by Bodigrim — APPROVED** [review]

**Comment by BebeSparkelSparkel** [comment]
Adding tests for lazy bytestring will small chunk size

File: benchmarks/haskell/Benchmarks/FileWrite.hs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        (h, removeFile) <- mkSinkNRemove
+        hSetBuffering h mode
+        hSetNewlineMode h $ NewlineMode nl nl
+        pure
+          ( bgroup (intercalate " " [textCharacteristics, show nl, show mode]) $
+            lengths <&> \n -> let
+              t = select $ L.toStrict $ L.take n writeData
+              in bench ("length " <> show n)
+                $ deepseq t
+                $ nfAppIO (hPutStr h) t
```
Comment (Lysxia) [inline benchmarks/haskell/Benchmarks/FileWrite.hs:51]: The result is `()` so `whnfAppIO` is enough.

**Comment by Bodigrim** [comment]
@Lysxia good to go?

**Review by Lysxia — APPROVED** [review]
Yes this is good to go.

I was just confused about how to interpret the results to justify an optimization to `putStr` in PR #589. See [this thread on an earlier version of it](https://github.com/haskell/text/pull/589#discussion_r1584358098)).

But these benchmarks are still useful in general to keep track of performance improvements to printing `Text`.

**Comment by Bodigrim** [comment]
Thanks @BebeSparkelSparkel!

**Pull request merged**